gmail/src/models/state.ts

```diff
@@ -0,0 +1,196 @@
+        this.error = error;
+    }
+
+    toJson(): string {
+        return JSON.stringify(this);
+    }
+
+    /**
+     * Unserialize the state object (reverse JSON.stringify).
+     */
+    static fromJson(json: string): State {
+        const values = JSON.parse(json);
+
+        const partnerValues = values.partner || {};
+        const emailValues = values.email || {};
+        const errorValues = values.error || {};
+        const partnersValues = values.partners;
+
+        const partner = Partner.fromJson(partnerValues);
+        const email = Email.fromJson(emailValues);
+        const error = ErrorMessage.fromJson(errorValues);
+        const partners = partnersValues
+            ? partnersValues.map((partnerValues: any) => Partner.fromJson(partnerValues))
+            : null;
+
+        return new State(partner, email, partners, error);
+    }
+
+    /**
+     * Cache / user properties management.
+     *
+     * Introduced with static getter / setter because they are shared between all the
+     * application cards.
+     */
+    static get accessToken() {
+        const accessToken = getAccessToken();
+        return this._isTrue(accessToken);
+    }
+
+    static get isLogged(): boolean {
+        return !!this.accessToken;
+    }
+
+    /**
+     * Return the URL require to login to the Odoo database.
+     */
+    static get odooLoginUrl(): string {
+        const loginUrl = getServiceUrl();
+        return this._isTrue(loginUrl);
+    }
+
+    /**
+     * Return the base URL of the Odoo database.
+     */
+    static get odooServerUrl(): string {
+        const userProperties = PropertiesService.getUserProperties();
+        const serverUrl = userProperties.getProperty("ODOO_SERVER_URL");
+        return this._isTrue(serverUrl);
+    }
+
+    /**
+     * Change the base URL of the Odoo database.
+     */
+    static set odooServerUrl(value: string) {
+        const userProperties = PropertiesService.getUserProperties();
+        userProperties.setProperty("ODOO_SERVER_URL", value);
+    }
+
+    /**
+     * Return the shared secret between the add-on and IAP
+     * (which is used to authenticate the add-on to IAP).
+     */
+    static get odooSharedSecret(): string {
+        const scriptProperties = PropertiesService.getScriptProperties();
+        const sharedSecret = scriptProperties.getProperty("ODOO_SHARED_SECRET");
+        return this._isTrue(sharedSecret);
+    }
+
+    /**
+     * Dictionary which inform us on which record we already logged the email.
+     * So the user can not log 2 times the same email on the same record.
+     * This is stored into the cache, so we don't need to modify the Odoo models.
+     *
+     * Note: the cache expire after 6 hours.
+     *
+     * Returns:
+     *     {
+     *         "partners": [3, 6], // email already logged on the partner 3 and 6
+     *         "leads": [7, 14],
+     *     }
+     */
+    static getLoggingState(messageId: string) {
```

Comment:
I thought we decided this was not necessary after all, if you log the email multiple times it doesn't break anything anyway?

Reply:
Yes, but it cost almost nothing to do it and LNA seems happy with this feature